napari_cellseg3d/code_models/models/wnet/soft_Ncuts.py

```diff
@@ -0,0 +1,225 @@
+
+    Args:
+        data_shape (H, W, D): shape of the images as a tuple.
+        intensity_sigma (scalar): scale of the gaussian kernel of pixels brightness.
+        spatial_sigma (scalar): scale of the gaussian kernel of pixels spacial distance.
+        radius (scalar): radius of pixels for which we compute the weights
+    """
+
+    def __init__(
+        self, data_shape, device, intensity_sigma, spatial_sigma, radius=None
+    ):
+        super(SoftNCutsLoss, self).__init__()
+        self.intensity_sigma = intensity_sigma
+        self.spatial_sigma = spatial_sigma
+        self.radius = radius
+        self.H = data_shape[0]
+        self.W = data_shape[1]
+        self.D = data_shape[2]
+        self.device = device
+
+        if self.radius is None:
+            self.radius = min(
+                max(5, math.ceil(min(self.H, self.W, self.D) / 20)),
+                self.H,
+                self.W,
+                self.D,
+            )
+        print(f"Radius set to {self.radius}")
+
+    def forward(self, labels, inputs):
+        """Forward pass of the Soft N-Cuts loss.
+
+        Args:
+            labels (torch.Tensor): Tensor of shape (N, K, H, W, D) containing the predicted class probabilities for each pixel.
+            inputs (torch.Tensor): Tensor of shape (N, C, H, W, D) containing the input images.
+
+        Returns:
+            The Soft N-Cuts loss of shape (N,).
+        """
+        # inputs.shape[0]
+        # inputs.shape[1]
+        K = labels.shape[1]
+
+        labels.to(self.device)
+        inputs.to(self.device)
+
+        loss = 0
+
+        kernel = self.gaussian_kernel(self.radius, self.spatial_sigma).to(
+            self.device
+        )
+
+        for k in range(K):
+            # Compute the average pixel value for this class, and the difference from each pixel
+            class_probs = labels[:, k].unsqueeze(1)
+            class_mean = torch.mean(
+                inputs * class_probs, dim=(2, 3, 4), keepdim=True
+            ) / torch.add(
+                torch.mean(class_probs, dim=(2, 3, 4), keepdim=True), 1e-5
+            )
+            diff = (inputs - class_mean).pow(2).sum(dim=1).unsqueeze(1)
+
+            # Weight the loss by the difference from the class average.
+            weights = torch.exp(
+                diff.pow(2).mul(-1 / self.intensity_sigma**2)
+            )
+
+            numerator = torch.sum(
+                class_probs
+                * F.conv3d(class_probs * weights, kernel, padding=self.radius),
+                dim=(1, 2, 3, 4),
+            )
+            denominator = torch.sum(
+                class_probs * F.conv3d(weights, kernel, padding=self.radius),
+                dim=(1, 2, 3, 4),
+            )
+            loss += nn.L1Loss()(
+                numerator / torch.add(denominator, 1e-6),
+                torch.zeros_like(numerator),
+            )
+
+        return K - loss
```

Comment:
wait, the overall loss is the number of classes - loss?

Reply:
Yes, that's why it goes up by 1 when you add classes. In our case it means that it's always bounded between 0 and 1

Reply:
It's due to needing the Soft version for backward prop